torch/csrc/distributed/autograd/functions/dist_accumulate_grad.cpp

```diff
@@ -0,0 +1,49 @@
+#include <torch/csrc/distributed/autograd/functions/dist_accumulate_grad.h>
+
+#include <torch/csrc/autograd/functions/accumulate_grad.h>
+#include <torch/csrc/autograd/functions/utils.h>
+#include <torch/csrc/autograd/variable.h>
+
+namespace torch::distributed::autograd {
+
+DistAccumulateGrad::DistAccumulateGrad(
+    AccumulateGrad&& accumulateGrad,
+    std::shared_ptr<DistAutogradContext> autogradContext)
+    : Node(accumulateGrad.sequence_nr()),
+      variable_(std::move(accumulateGrad.variable)),
+      autogradContext_(std::move(autogradContext)) {
+  move_from(std::move(accumulateGrad));
```

Comment:
I don't think you want to steal stuff from the AccumulateGrad. The AccumulateGrad node can outlive this single backward call and might be used again.

Reply:
The `AccumulateGrad` node is not a part of the autograd graph. How can others refer to it?
(though I know DDP reducers, which create post hooks, keep references to them)

https://github.com/pytorch/pytorch/blob/15a7b9cf0ab4674e798d735042cdcc9184b66e8a/torch/csrc/distributed/c10d/reducer.cpp#L195-L207

Reply:
It is actually in the graph. This is why you get it by looking into `next_functions`. Being in the graph is actually the only thing that keeps it alive as, as you said, the variable itself only has a weak_ref to it.

Reply:
`DistAccumulateGrad` is a replacement of `AccumulateGrad` in the graph. Here it has to steal metadata, most of which are in `unique_ptr`, form `AccumulateGrad`. I know `AccumulateGrad` nodes can outlive the graph. One example is DDP reducer keeps `shared_ptr` to them.

Reply:
But what happens if someone does:
```
inp = torch.rand(10, requires_grad=True)
loss = model(inp)
with dist_autograd.context() as context_id:
  dist_autograd.backward([loss], retain_graph=True)
loss.backward()
```

I don't claim I have a use case for this. But if we don't support this, it should be explicitly checked and should raise a nice error.

Reply:
For existing DDP, calling backward twice would crash, as it requires a forward pass to mark `expect_autograd_hooks_` to `true`.

Reply:
> But what happens if someone does:
> 
> ```
> inp = torch.rand(10, requires_grad=True)
> loss = model(inp)
> with dist_autograd.context() as context_id:
>   dist_autograd.backward([loss], retain_graph=True)
> loss.backward()
> ```
> 
> I don't claim I have a use case for this. But if we don't support this, it should be explicitly checked and should raise a nice error.

I'll make this one a test case.